gap/SnAnUnknownDegree.gi

```diff
@@ -0,0 +1,710 @@
+#
+# TODO: take care of duplicate candidates?
+# Creates and returns a function, here called oneThreeCycleCandidate. The
+# function oneThreeCycleCandidate returns one of the following:
+# - a three cycle candidate, i.e. an element of G
+# - TemporaryFailure, if we exhausted all attempts
+# - NeverApplicable, if we found out that G can't be an Sn or An
+BindGlobal("ThreeCycleCandidatesIterator",
+    function(ri, eps, N)
+    local
+        # involution
+        t,
+        # integers, controlling the number of iterations
+        M, B, T, C, logInt2N,
+        # iteration over all primes up to N
+        allPrimes, i, p,
+        # counters
+        nrInvolutions, nrTriedConjugates, nrThreeCycleCandidates,
+        # helper functions
+        tryThreeCycleCandidate, oneThreeCycleCandidate;
+    # Step 1: Initialization
+    # The current involution t_i
+    t := fail;
+
+    # Constants
+    # TODO: better iteration over primes
+    M := 1;
+    allPrimes := AllPrimesUpTo(N);
+    for i in [2 .. Length(allPrimes)] do
+        p := allPrimes[i];
+        M := M * p ^ LogInt(N, p);
+    od;
+    # FIXME: Probably B can be chosen smaller
+    B := Int(Ceil(13 * Log2(Float(N)) * Log2(3 / Float(eps))));
+    T := Int(Ceil(3 * Log2(3 / Float(eps))));
+    C := Int(Ceil(Float(3 * N * T / 5)));
+    logInt2N := LogInt(N, 2);
+
+    # Counters
+    # Counts the constructed involutions t_i in steps 2 & 3.
+    nrInvolutions := 0;
+    # Counts the elements c in step 4 that we use to conjugate the current
+    # involution t_i.  We initialize nrTriedConjugates to C such that "steps 2
+    # & 3" in tryThreeCycleCandidate immediately construct an involution.
+    nrTriedConjugates := C;
+    # counts the size of the set Gamma_i in step 4 for the current involution
+    # t_i
+    nrThreeCycleCandidates := 0;
+
+    # Helper functions
+    # tryThreeCycleCandidate returns one of the following:
+    # - a three cycle candidate, i.e. an element of G
+    # - fail, if the random conjugate c from step 4 and t commute. Then we have
+    #   to call tryThreeCycleCandidate again
+    # - NeverApplicable, if G can not be an Sn or An
+    tryThreeCycleCandidate := function()
+        local
+            # integer, loop variable
+            a,
+            # elements, in G
+            r, tPower, tPowerOld, c;
+        # Steps 2 & 3: New involution
+        # Check if we either tried enough conjugates or constructed enough
+        # three cycle candidates for the current involution t.
+        # If this is the case, we need to construct the next involution
+        if nrTriedConjugates >= C or nrThreeCycleCandidates >= T then
+            r := RandomElm(ri, "simplesocle", true)!.el;
+            a := 0;
+            tPower := r ^ M;
+            # Invariant: tPower = (r ^ M) ^ (2 ^ a)
+            repeat
+                a := a + 1;
+                tPowerOld := tPower;
+                tPower := tPower ^ 2;
+            until a = logInt2N or isone(ri)(tPower);
```

Comment:
Minor tweak, since for loops are optimized in GAP:
```suggestion
            tPower := r ^ M;
            # Invariant: tPower = (r ^ M) ^ (2 ^ a)
            for a in [1..logInt2N]
                tPowerOld := tPower;
                tPower := tPower ^ 2;
                if isone(ri)(tPower) then break; fi;
            od;
```